File: longest-palindrome-by-concatenating-two-letter-words.rs
use std::collections::HashMap;

impl Solution {
  pub fn longest_palindrome(words: Vec<String>) -> i32 {
    let mut freq: HashMap<String, i32> = words
            .into_iter()
            .fold(HashMap::new(), |mut f, w| { *f.entry(w).or_default() += 1; f });

    let (mut o, mut ans) = (false, 0);
    for (word, count) in freq.iter() {
      if word.bytes().rev().eq(word.bytes()) {
        ans += count / 2 * 2;
        if count % 2 == 1 {
          o = true;
        }
      }
      else {
        let rstr = word.chars().rev().collect::<String>();
        ans += count.min(freq.get(&rstr).unwrap_or(&0));
      }
    }

    if o { ans += 1; }
    ans * 2
  }
}

File: l/longest-palindrome-by-concatenating-two-letter-words.cc
class Solution {
public:
  int longestPalindrome(vector<string>& words) {
    unordered_map<string, int> tal;
    for (auto& word : words) ++tal[word];

    int ans = 0, o = 0;
    for (char c1 = 'a'; c1 <= 'z'; ++c1) for (char c2 = c1 + 1; c2 <= 'z'; ++c2) {
      string s1(1, c1); s1 += c2;
      string s2(1, c2); s2 += c1;
      ans += min(tal[s1], tal[s2]) * 4;
    }
    for (char c = 'a'; c <= 'z'; ++c) {
      string s(2, c);
      ans += (tal[s] / 2) * 4;
      if (tal[s] % 2) o = 2;
    }
    return ans + o;
  }
};
